python_modules/dagster/dagster/core/definitions/assets_info.py

```diff
@@ -0,0 +1,183 @@
+    ):
+        return super().__new__(
+            cls,
+            asset_key=check.inst_param(asset_key, "asset_key", AssetKey),
+            asset_partitions_fn=check.opt_callable_param(
+                asset_partitions_fn, "asset_partitions_fn", lambda _: None
+            ),
+            asset_partitions_def=check.opt_inst_param(
+                asset_partitions_def, "asset_partitions_def", PartitionsDefinition
+            ),
+        )
+
+
+def _assets_job_info_for_node(
+    node_def: NodeDefinition, node_handle: Optional[NodeHandle]
+) -> Tuple[
+    Mapping[NodeInputHandle, AssetKey],
+    Mapping[NodeOutputHandle, AssetOutputInfo],
+    Mapping[AssetKey, AbstractSet[AssetKey]],
+]:
+    """
+    Recursively iterate through all the sub-nodes of a Node to find any ops with asset info
+    encoded on their inputs/outputs
+    """
+    check.inst_param(node_def, "node_def", NodeDefinition)
+    check.opt_inst_param(node_handle, "node_handle", NodeHandle)
+
+    asset_key_by_input: Dict[NodeInputHandle, AssetKey] = {}
+    asset_info_by_output: Dict[NodeOutputHandle, AssetOutputInfo] = {}
+    asset_deps: Dict[AssetKey, AbstractSet[AssetKey]] = {}
+    if not isinstance(node_def, GraphDefinition):
+        # must be in an op (or solid)
+        if node_handle is None:
+            check.failed("Must have node_handle for non-graph NodeDefinition")
+        input_asset_keys: Set[AssetKey] = set()
+        for input_def in node_def.input_defs:
+            input_key = input_def.hardcoded_asset_key
+            if input_key:
+                input_asset_keys.add(input_key)
+                input_handle = NodeInputHandle(node_handle, input_def.name)
+                asset_key_by_input[input_handle] = input_key
+        for output_def in node_def.output_defs:
+            output_key = output_def.hardcoded_asset_key
+            if output_key:
+                output_handle = NodeOutputHandle(node_handle, output_def.name)
+                asset_info_by_output[output_handle] = AssetOutputInfo(
+                    asset_key=output_key,
+                    asset_partitions_fn=output_def.get_asset_partitions,
+                    asset_partitions_def=output_def.asset_partitions_def,
+                )
+                # assume output depends on all inputs
+                asset_deps[output_key] = input_asset_keys
+    else:
+        # keep recursing through structure
+        for sub_node_name, sub_node in node_def.node_dict.items():
+            n_asset_key_by_input, n_asset_info_by_output, n_asset_deps = _assets_job_info_for_node(
+                node_def=sub_node.definition,
+                node_handle=NodeHandle(sub_node_name, parent=node_handle),
+            )
+            asset_key_by_input.update(n_asset_key_by_input)
+            asset_info_by_output.update(n_asset_info_by_output)
+            asset_deps.update(n_asset_deps)
+    return asset_key_by_input, asset_info_by_output, asset_deps
+
+
+class AssetsJobInfo:
```

Comment:
riffing on names:
* `AssetsMappings`
* `AssetMapping`
* `AssetLayer`
* `AssetResolution`
* `AssetResolver` (maybe better if fns for processing were on obj)
* `AssetConstraints`
* `AssetStuff`

Reply:
`AssetStuff` wins my heart, but `AssetLayer` feels surprisingly nice to me.

job.asset_layer.asset_key_for_output -> "I'm querying the Job's asset layer for an output".

`AssetsMappings` is maybe the most accurate for the current state of the object (it's a bunch of dictionaries, after all), but is potentially more limiting for future extensions to this object.

Happy with either of those two, and for now I'll switch to AssetLayer as I mess with this stack.